### PR TITLE
Restart dynamo analysis when we fail to tensorify away all symfloat inputs

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140893
* #138922
* #138915
* __->__ #140346

Fixes a bunch of benchmarks that failed with cudagraph errors including `tlp python benchmarks/dynamo/timm_models.py --device cuda --inductor --accuracy --amp --training --only resmlp_12_224` when `specialize_float=False`

Also brings down number of overall failures (with keep-going) from 108 => 62. I'd estimate >80% of those 62 are wobbly expect tests. 

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 1]
```diff
@@ -281,6 +283,21 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                     node.replace_all_uses_with(guard_scalar(val))
                     graph.erase_node(node)
 
+    should_restart = False
+    for i, node in enumerate(graph.nodes):
+        if node.op == "placeholder" and len(node.users) == 0:
+            # At this point we've lost the back pointer to
+            # what f_local this placeholder points to. Instead,
+            # we will rely on the index to specialize when we
+            # restart analysis.
```
Comment (bobrenjc93) [inline torch/fx/passes/_tensorify_python_scalars.py:292]: This is probably the most controversial part of this PR. An alternative universe is to have AOTAutoGrad propogate placeholder names to this point and we can have hoist the specialization logic up to the wrap_symfloat level.

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 2]
```diff
@@ -281,6 +283,21 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                     node.replace_all_uses_with(guard_scalar(val))
                     graph.erase_node(node)
 
+    should_restart = False
+    for i, node in enumerate(graph.nodes):
+        if node.op == "placeholder" and len(node.users) == 0:
```
Comment (bobrenjc93) [inline torch/fx/passes/_tensorify_python_scalars.py:288]: If we want to optimize for performance perhaps it make sense to break early once we iterate past the placeholders.

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 3]
```diff
@@ -281,6 +288,37 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
...
+            and (name := placeholder_to_symfloat_name.get(node))
+        ):
+            # At this point we've lost the back pointer to
+            # what f_local this placeholder points to. Instead,
+            # we will rely on the symfloat name to specialize when we
+            # restart analysis.
+            TensorifyState.specialize(name)
+            should_restart = True
+
+    # Sometimes by the time we get to tensorify, there have already been
```
Comment (bobrenjc93) [inline torch/fx/passes/_tensorify_python_scalars.py:305]: This is the meat of the PR. Because by the time we reach tensorify we've lost all source/name information in the fx graph, I utilize shape_env to figure out which specializations have already happened.

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 4]
```diff
@@ -281,6 +307,48 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
...
+    # using node meta val. Here's an example graph where you have no replacements but
+    # still want to specialize and get rid of zf0 before inductor:
+    #
+    # def forward(self, L_tensor_: "f16[2, 3, 4][12, 4, 1]cuda:0", L_scale_factor_: "f64[][]cpu"):
+    #     l_tensor_ = L_tensor_
+    #     l_scale_factor_ = L_scale_factor_
+    #     item: "Sym(zf0)" = l_scale_factor_.item();  l_scale_factor_ = None
+    #     interpolate: "f16[2, 3, TruncToInt(4.0*zf0)][3*TruncToInt(4.0*zf0), TruncToInt(4.0*zf0), 1]cuda:0" = torch.nn.functional.interpolate(l_tensor_, size = None, scale_factor = item, mode = 'nearest-exact', align_corners = None, recompute_scale_factor = True);  l_tensor_ = item = None # noqa: B950
+    #     return (interpolate,)
+    for i, node in enumerate(graph.nodes):
```
Comment (bobrenjc93) [inline torch/fx/passes/_tensorify_python_scalars.py:321]: I have an experimental improvement in https://github.com/pytorch/pytorch/pull/140829

I'll wait for CI to get green before squashing but the high level idea is to instead maintain a whitelist of symfloats we managed to tensorify away and then specialize anything not in the whitelist. Makes a lot of this logic much cleaner.

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66032797).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66032986).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66032986).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66033375).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66033375).

File: test/dynamo/test_unspec.py [hunk 5]
```diff
@@ -659,7 +659,7 @@ def fn(x, y):
 
             cnt = CompileCounterWithBackend("aot_eager")
             fn_opt = torch._dynamo.optimize(cnt)(fn)
-            x = torch.tensor(9.734375, dtype=dtype, requires_grad=True)
+            x = torch.randn(5, dtype=dtype, requires_grad=True)
```
Comment (bobrenjc93) [inline test/dynamo/test_unspec.py:662]: I just changed this to be a non scalar tensor to make the test simpler, otherwise we would have different frame counts depending on if dtype == torch.float64. This is because were were assigning zf0 to 9.734375 and zf1 to the concrete value of y. So at the time of tensorify, we would tensorify away zf1 but not zf0 causing a restart.

I suppose this highlights one of the downsides of the current implementation where sometimes we will restart analysis when we don't really need to. I guess in the most ideal world at the time of tensorify we would know what was specialized before this point in time (eg. in arg_parser.h or in fake tensor prop). Then we could know which non tensorified away symfloats are okay.

File: torch/_dynamo/convert_frame.py [hunk 6]
```diff
@@ -758,6 +765,8 @@ def log_bytecode(
                 if attempt > 100:
                     unimplemented("100+ RestartAnalysis() calls")
             except exc.SkipFrame as e:
+                if not isinstance(e, exc.TensorifyScalarRestartAnalysis):
+                    TensorifyState.clear()
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:769]: I'm not sure how to refactor this but having to manually clear the states here seems error prone (and I also remember stubbing my toe on this with compiler collectives, see https://github.com/pytorch/pytorch/pull/131983

File: torch/_dynamo/symbolic_convert.py [hunk 7]
```diff
@@ -245,6 +246,22 @@ class DistributedState:
     all_states: Optional[List[LocalState]] = None
 
 
+class TensorifyState:
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:249]: Docs. Esp what the str denotes here

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 8]
```diff
@@ -281,6 +306,27 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
...
+                not TensorifyState.should_specialize(name)
+                and name not in tensorified_symbols
+            ):
+                TensorifyState.specialize(name)
+                should_restart = True
+
+    if should_restart:
+        # Sledgehammer time. Restart dynamo analysis, keeping track of which input sources
+        # are no longer needed and should be specialized.
+        raise TensorifyScalarRestartAnalysis
```
Comment (ezyang) [inline torch/fx/passes/_tensorify_python_scalars.py:328]: Similar to https://github.com/pytorch/pytorch/pull/131983 need to clear speculation log

Comment (ezyang) [reply torch/fx/passes/_tensorify_python_scalars.py:328]: Not sure what happens if compiler collectives is going on at this point too

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 9]
```diff
@@ -109,6 +112,8 @@ def tensorify_python_scalars(
     tracer = fx.proxy.GraphAppendingTracer(graph)
     expr_to_sym_proxy: dict[sympy.Expr, MetaProxy] = {}
     expr_to_tensor_proxy: dict[sympy.Expr, MetaProxy] = {}
+    tensorified_symbols: set[str] = set()
```
Comment (ezyang) [inline torch/fx/passes/_tensorify_python_scalars.py:115]: Why not just store sympy.Symbol here, if it's symbols

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 10]
```diff
@@ -207,6 +211,24 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                         node, tracer=tracer, fake_mode=fake_mode
                     )
 
+            # Specialize all dimensions that contain symfloats. Here's
+            # an example test that requires this:
+            # PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=4 tlp python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_nn_functional_interpolate_bicubic_cuda_float32 # noqa: B950
```
Comment (ezyang) [inline torch/fx/passes/_tensorify_python_scalars.py:216]: s/tlp//

Comment (ezyang) [reply torch/fx/passes/_tensorify_python_scalars.py:216]: This seems incredibly suspicious. You shouldn't ever have a SymFloat that comes out of tensor.shape....

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 11]
```diff
@@ -207,6 +211,24 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
...
+            val = node.meta.get("val")
+            if isinstance(val, FakeTensor):
+                for dim in val.shape:
+                    if isinstance(
+                        dim,
+                        (torch.SymFloat, torch.SymInt, torch.SymBool),
```
Comment (ezyang) [inline torch/fx/passes/_tensorify_python_scalars.py:222]: so it seems to me that the SymInt test here should suffice

File: torch/fx/passes/_tensorify_python_scalars.py (hunk 10 above)
Comment (bobrenjc93) [reply torch/fx/passes/_tensorify_python_scalars.py:216]: IIRC it was something like Floor(zf0 * 1.5) or something like that. I can run the repro again to get the exact condition if you'd like.

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 12]
```diff
@@ -207,6 +211,24 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
...
+                    ):
+                        for s in dim.node.expr.free_symbols:
+                            name = str(s)
+                            if symbol_is_type(
+                                s, SymT.FLOAT
+                            ) and not TensorifyState.should_specialize(name):
+                                TensorifyState.specialize(name)
+                                should_restart = True
```
Comment (ezyang) [inline torch/fx/passes/_tensorify_python_scalars.py:230]: This is good, although it's probably worth elaborating in a little more detail why we choose to spec here. In principle, we could support float input that is used to do size compute. The problem is that we don't actually want to tensorify the compute in this case, which means need support for codegen all the float! (Maybe if someone comes to us with a legitimate use case, this restriction could be lifted, perhaps.)

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 13]
```diff
@@ -281,6 +306,27 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                     node.replace_all_uses_with(guard_scalar(val))
                     graph.erase_node(node)
 
+    # Sometimes by the time we get to tensorify, there have already been
+    # specializations, eg. in python_arg_parser.h. In these cases,
+    # placeholder nodes no longer have a reference to their original
+    # symfloat and thus we need to deduce specializations have happend
+    # via shape_env.replacements. NB: there's an important invariant here
+    # that symfloats keep consistent names across restarts.
```
Comment (ezyang) [inline torch/fx/passes/_tensorify_python_scalars.py:314]: So what, you have to allocate the symfloat name even if you're not going to use it at all? Seems sus lol

Comment (ezyang) [reply torch/fx/passes/_tensorify_python_scalars.py:314]: Don't we associate every var with sources?

```
        self.var_to_sources: Dict[sympy.Symbol, List[Source]] = {}
```

So that's how you get sources.

File: torch/fx/passes/_tensorify_python_scalars.py (hunk 8 above)
Comment (ezyang) [reply torch/fx/passes/_tensorify_python_scalars.py:328]: I'm looking for a comment that says something like "restarting analysis is necessary because we need to instruct Dynamo to NOT make these as inputs."

It really does seem like CUDA graph should be able to ignore inputs that don't get used inside the body though LOL

File: torch/_dynamo/output_graph.py [hunk 14]
```diff
@@ -1617,6 +1628,38 @@ def update_used_symbols(used_symbols, fake: Union[torch.SymInt, torch.Tensor]):
...
+        # This is a pretty interesting function. Basically we have this problem
+        # where our compiler tends to choke when we have unused inputs. The way
+        # we support dynamic float arguments is by doing a joint fx pass and
+        # tensorifying away as many symfloats as we can. For the remaining symfloats
+        # we have no choice but to specialize... HOWEVER at that point in time
+        # we can no longer remove graph inputs. So our sledgehammer solution is to
+        # save the state of what inputs we should have specialized in dynamo and
+        # restart analysis. This function incorporates this "view from the future"
+        # state and specializes inputs that we know we won't be able to tensorify
+        # away in the joint pass.
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:1641]: So explicitly say that in principle we shouldn't choke and so this is not necessary

File: torch/_dynamo/output_graph.py [hunk 15]
```diff
@@ -1617,6 +1628,38 @@ def update_used_symbols(used_symbols, fake: Union[torch.SymInt, torch.Tensor]):
...
+
+        # Import here to prevent circular import
+        from torch._dynamo.symbolic_convert import TensorifyState
+
+        for node in self.graph.nodes:
+            example_value = node.meta.get("example_value")
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:1647]: I feel like grapharg should more directly get you to what you want here

**Review by ezyang — APPROVED** [review]
Looks good, please see comments

File: torch/fx/passes/_tensorify_python_scalars.py (hunk 10 above)
Comment (ezyang) [reply torch/fx/passes/_tensorify_python_scalars.py:216]: Definitely a bug. FloorToInt should be used when it gets coerced to int, if that's the case.
